File: 백준/Silver/2468. 안전 영역/안전 영역.py
from collections import deque

def bfs(ny,nx,nk):
    global visit
    diy = [0,0,1,-1]
    dix = [1,-1,0,0]
    q = deque()
    q.append([ny,nx])
    while q:
        y,x = q.popleft()
        for l in range(4):
            dy = diy[l]+y
            dx = dix[l]+x
            if dy<0 or dy>=N or dx<0 or dx>=N:
                continue
            if arr[dy][dx] <=nk:
                continue
            if visit[dy][dx] == 1:
                continue
            visit[dy][dx] = 1
            q.append([dy,dx])


N = int(input())
MAX = -21e8
arr = []
for _ in range(N):
    ar = list(map(int,input().split()))
    if MAX<max(ar):
        MAX = max(ar)
    arr.append(ar)

lst = [0]*(MAX+1)
for i in arr:
    for j in i:
        lst[j] = 1

ans = [0] *(MAX+1)
for k in range(MAX+1):
    if lst[k] == 1:
        visit = [[0]*N for _ in range(N)]
        cnt = 0
        for i in range(N):
            for j in range(N):
                if visit[i][j] == 0 and arr[i][j]>k:
                    visit[i][j] = 1
                    bfs(i,j,k)
                    cnt += 1
        ans[k] += cnt
if max(ans) == 0:
    print(1)
else:
    print(max(ans))

File: 백준/Silver/2468. 안전 영역/안전 영역.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    static int N;
    static int[][] arr;
    static int[][] visit;
    static int[] diy = {0, 0, 1, -1};
    static int[] dix = {1, -1, 0, 0};

    public static void bfs(int ny, int nx, int nk) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{ny, nx});
        
        while (!q.isEmpty()) {
            int[] current = q.poll();
            int y = current[0];
            int x = current[1];

            for (int l = 0; l < 4; l++) {
                int dy = diy[l] + y;
                int dx = dix[l] + x;

                if (dy < 0 || dy >= N || dx < 0 || dx >= N) {
                    continue;
                }
                if (arr[dy][dx] <= nk) {
                    continue;
                }
                if (visit[dy][dx] == 1) {
                    continue;
                }
                visit[dy][dx] = 1;
                q.offer(new int[]{dy, dx});
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        int MAX = 0;
        arr = new int[N][N];

        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                arr[i][j] = sc.nextInt();
                if (arr[i][j] > MAX) {
                    MAX = arr[i][j];
                }
            }
        }

        int[] ans = new int[MAX + 1];
        
        for (int k = 0; k <= MAX; k++) {
            visit = new int[N][N];
            int cnt = 0;
            for (int i = 0; i < N; i++) {
                for (int j = 0; j < N; j++) {
                    if (visit[i][j] == 0 && arr[i][j] > k) {
                        visit[i][j] = 1;
                        bfs(i, j, k);
                        cnt++;
                    }
                }
            }
            ans[k] = cnt;
        }

        int maxAns = 0;
        for (int count : ans) {
            if (count > maxAns) {
                maxAns = count;
            }
        }

        if (maxAns == 0) {
            System.out.println(1);
        } else {
            System.out.println(maxAns);
        }
    }
}